lucene/core/src/java/org/apache/lucene/store/NRTCachingDirectory.java

```diff
@@ -0,0 +1,324 @@
+    }
+    in.sync(fileNames);
+  }
+
+  @Override
+  public void rename(String source, String dest) throws IOException {
+    unCache(source);
+    if (cacheDirectory.fileExists(dest)) {
+      throw new IllegalArgumentException("target file " + dest + " already exists");
+    }
+    in.rename(source, dest);
+  }
+
+  @Override
+  public synchronized IndexInput openInput(String name, IOContext context) throws IOException {
+    if (VERBOSE) {
+      System.out.println("nrtdir.openInput name=" + name);
+    }
+    if (cacheDirectory.fileExists(name)) {
+      if (VERBOSE) {
+        System.out.println("  from cache");
+      }
+      return cacheDirectory.openInput(name, context);
+    } else {
+      return in.openInput(name, context);
+    }
+  }
+  
+  /** Close this directory, which flushes any cached files
+   *  to the delegate and then closes the delegate. */
+  @Override
+  public void close() throws IOException {
+    // NOTE: technically we shouldn't have to do this, ie,
+    // IndexWriter should have sync'd all files, but we do
+    // it for defensive reasons... or in case the app is
+    // doing something custom (creating outputs directly w/o
+    // using IndexWriter):
+    IOUtils.close(
+        () -> {
+          if (!closed.getAndSet(true)) {
+            for(String fileName : cacheDirectory.listAll()) {
+              unCache(fileName);
+            }
+          }
+        },
+        cacheDirectory,
+        in);
+  }
+
+  /** Subclass can override this to customize logic; return
+   *  true if this file should be written to the RAM-based cache first. */
+  protected boolean doCacheWrite(String name, IOContext context) {
+    //System.out.println(Thread.currentThread().getName() + ": CACHE check merge=" + merge + " size=" + (merge==null ? 0 : merge.estimatedMergeBytes));
+
+    long bytes = 0;
+    if (context.mergeInfo != null) {
+      bytes = context.mergeInfo.estimatedMergeBytes;
+    } else if (context.flushInfo != null) {
+      bytes = context.flushInfo.estimatedSegmentSize;
+    } else {
+      return false;
+    }
+
+    return (bytes <= maxMergeSizeBytes) && (bytes + cacheSize.get()) <= maxCachedBytes;
+  }
+
+  @Override
+  public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
+    if (VERBOSE) {
+      System.out.println("nrtdir.createTempOutput prefix=" + prefix + " suffix=" + suffix);
+    }
+    Set<String> toDelete = new HashSet<>();
+
+    // This is very ugly/messy/dangerous (can in some disastrous case maybe create too many temp files), but I don't know of a cleaner way:
+    boolean success = false;
+
+    Directory first;
+    Directory second;
+    if (doCacheWrite(prefix, context)) {
+      first = cacheDirectory;
+      second = in;
+    } else {
+      first = in;
+      second = cacheDirectory;
+    }
+
+    IndexOutput out = null;
+    try {
+      while (true) {
+        out = first.createTempOutput(prefix, suffix, context);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *out*. The resource is referenced in statements at the following lines: 262, 266, 273. The resource closure statement is at lines: 269, 280. The resource is returned at line: 285. There are other execution paths that neither close the resource nor return it, e.g., when exception is thrown by IOUtils.deleteFiles, and if-branch is taken at line 277, and else-branch is taken at line 268. To prevent this resource leak, close **out** along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).